### ISSUE TITLE
Do I need two different servers to judge my baseline and canary values??

### ISSUE THREAD

**Issue by amsin02** [spinnaker/kayenta#540]
I am currently using **two different ports** to run my Canary & Baseline applications on the **same server i.e . localhost** . 
But, Kayenta-standalone-canary-analysis is taking identical values for canary and baseline for the asked metric.

**It takes same values for canary and baseline from my App1 (Canary) and 
again takes both canary and baseline values from App2 (Baseline)**.
 I want it to take canary values from App1 and baseline values from App2 on same server. 

I am using Prometheus as my metric source & publishing same metricName with different labels from both my applications i.e.
```
App1--> num_of_error requests{scope="canary", instance="localhost:8019"}
App1--> num_of_error requests{scope="baseline", instance="localhost:8020"} 
```
Below is my executionRequest:

```"canaryAnalysisExecutionRequest": {
    "scopes": [
      {
        "scopeName": "default",
        "controlScope": "localhost:8019",   // i have tried with controlScope="canary"
        "experimentScope": "localhost:8020",  // and setting  experimentScope="baseline" also. 
        "step": 15
      }
    ],
    "thresholds": {
      "pass": 75,
      "marginal": 95
    },
    "lifetimeDurationMins": 1,
    "beginAfterMins": 0,
    "lookbackMins": 0,
    "analysisIntervalMins": 1,
    "siteLocal": {}
  } 
```

**Comment by fieldju** [comment]
I am going to use SignalFx and AWS as an example because that is what I use.

The tl;dr is no, you don't need separate servers.

Kayenta doesn't care or know about your infrastructure. The only thing that matters is that you are reporting metrics to a metric source in a way that Kayenta can make queries to get results for some metric over some time period for the experiment and baseline separately.

Some background:
- SignalFx is a dimension aware metric source. 
- Kayenta wants a scope / location for the baseline and experiment
  - `Scope`: This should be the [server_group](https://www.spinnaker.io/concepts/#server-group) name. It is a unique identifier for the `experiment` / `control` / `prod` / `deployed unit`
    - In my case I add a dimension called `server_group` to my SignalFx metrics reporter and populate with it with the value of the Auto Scaling Group (ASG). ex: `cerberus-prod-experiment-v1` or `cerberus-prod-baseline-v1`
  - `Location`: We add a dimension called `server_region` and populate it with the AWS Region.

For the SignalFx integration with Kayenta you setup what the dimension keys are for server group and location. So when we add a SignalFx account to our Kayenta it is like so

```yaml
kayenta:
  signalfx:
    enabled: true
    accounts:
    - name: sfx-architecture
      accessToken: xxxxxxxxxxxxxx
      defaultScopeKey: server_group
      defualtLocationKey: server_region
      supportedTypes:
      - METRICS_STORE
```

Now when I call `/canary` or `/standalone_canary_analysis` I use the server_group and server_region values in the call. These end up being wired into the query that is made to SignalFx.

ex:

If I make a call to `/standalone_canary_analysis` with the following `canaryAnalysisExecutionRequest` (Ingore the data in siteLocal that is for some custom event listeners that are not upstream)

```json
{  
  "canaryAnalysisExecutionRequest": {
    "scopes": [
      {
        "scopeName": "default",
        "controlScope": "carebears-cms-master-cerberus-prod2-baseline-v1",
        "controlLocation": "us-west-2",
        "experimentScope": "carebears-cms-master-cerberus-prod2-experiment-v1",
        "experimentLocation": "us-west-2",
        "step": 10
      }
    ],
    "thresholds": {
      "pass": 95,
      "marginal": 75
    },
    "lifetimeDurationMins": 45,
    "beginAfterMins": 1,
    "lookbackMins": 0,
    "analysisIntervalMins": 15,
    "siteLocal": {
      "applicationMetadata": {
        "environment": "prod-west",
        "domain": "platform tools",
        "department": "arch - cloud acceleration"
      },
      "notificationChannels": [
        {
          "destination": "cerberus-logs",
          "type": "slack",
          "on": [
            "success"
          ]
        },
        {
          "destination": "ca-zookeepers",
          "type": "slack",
          "on": [
            "failure"
          ]
        },
        {
          "destination": "15555555555",
          "type": "sms",
          "on": [
            "failure",
            "success"
          ]
        }
      ],
      "jenkinsMetadata": {
        "canaryMode": "manual",
        "buildUrl": "https://xxxxx.example.com/job/cerberus-management-service/job/cerberus-management-servce-ec2-bg-pipeline/job/master/107/"
      }
    }
  },
  "canaryConfig": {
    "name": "cerberus-canary-config",
    "description": "Cerberus Management Service Canary Configuration",
    "configVersion": "1",
    "applications": [
      "cerberus management service"
    ],
    "judge": {
      "name": "NetflixACAJudge-v1.0",
      "judgeConfigurations": {}
    },
    "metrics": [
      {
        "name": "Average Endpoint Latency",
        "query": {
          "type": "signalfx",
          "metricName": "requests.mean",
          "queryPairs": [],
          "aggregationMethod": "mean",
          "serviceType": "signalfx"
        },
        "groups": [
          "Latency"
        ],
        "analysisConfigurations": {
          "canary": {
            "critical": false,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "Bad Request Count",
        "query": {
          "type": "signalfx",
          "metricName": "requests.count",
          "queryPairs": [
            {
              "key": "response_code",
              "value": "4*"
            }
          ],
          "aggregationMethod": "sum",
          "serviceType": "signalfx"
        },
        "groups": [
          "Errors"
        ],
        "analysisConfigurations": {
          "canary": {
            "nanStrategy": "replace",
            "critical": false,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "CPU Usage",
        "query": {
          "type": "signalfx",
          "metricName": "cpu.utilization",
          "queryPairs": [],
          "aggregationMethod": "mean",
          "serviceType": "signalfx"
        },
        "groups": [
          "Saturation"
        ],
        "analysisConfigurations": {
          "canary": {
            "nanStrategy": "remove",
            "critical": false,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "JVM Memory Heap used",
        "query": {
          "type": "signalfx",
          "metricName": "jmx_memory.used",
          "queryPairs": [
            {
              "key": "plugin_instance",
              "value": "memory-heap"
            }
          ],
          "aggregationMethod": "mean",
          "serviceType": "signalfx"
        },
        "groups": [
          "Saturation"
        ],
        "analysisConfigurations": {
          "canary": {
            "nanStrategy": "remove",
            "critical": false,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "KMS Auth Latency",
        "query": {
          "type": "signalfx",
          "metricName": "requests.mean",
          "queryPairs": [
            {
              "key": "uri",
              "value": "/v2/auth/iam-principal"
            },
            {
              "key": "method",
              "value": "POST"
            },
            {
              "key": "response_code",
              "value": "200"
            }
          ],
          "aggregationMethod": "mean",
          "serviceType": "signalfx"
        },
        "groups": [
          "Latency"
        ],
        "analysisConfigurations": {
          "canary": {
            "critical": false,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "Memory Usage",
        "query": {
          "type": "signalfx",
          "metricName": "memory.utilization",
          "queryPairs": [
            {
              "key": "plugin",
              "value": "signalfx-metadata"
            }
          ],
          "aggregationMethod": "mean",
          "serviceType": "signalfx"
        },
        "groups": [
          "Saturation"
        ],
        "analysisConfigurations": {
          "canary": {
            "nanStrategy": "remove",
            "critical": false,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "Secure Data Read Latency",
        "query": {
          "type": "signalfx",
          "metricName": "requests.mean",
          "queryPairs": [
            {
              "key": "uri",
              "value": "/v1/secret/**"
            },
            {
              "key": "method",
              "value": "GET"
            },
            {
              "key": "response_code",
              "value": "200"
            }
          ],
          "aggregationMethod": "mean",
          "serviceType": "signalfx"
        },
        "groups": [
          "Latency"
        ],
        "analysisConfigurations": {
          "canary": {
            "critical": false,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "Server Error Count",
        "query": {
          "type": "signalfx",
          "metricName": "requests.count",
          "queryPairs": [
            {
              "key": "response_code",
              "value": "5*"
            }
          ],
          "aggregationMethod": "sum",
          "serviceType": "signalfx"
        },
        "groups": [
          "Errors"
        ],
        "analysisConfigurations": {
          "canary": {
            "nanStrategy": "replace",
            "critical": true,
            "direction": "increase"
          }
        },
        "scopeName": "default"
      },
      {
        "name": "Success Count",
        "query": {
          "type": "signalfx",
          "metricName": "requests.count",
          "queryPairs": [
            {
              "key": "response_code",
              "value": "2*"
            }
          ],
          "aggregationMethod": "sum",
          "serviceType": "signalfx"
        },
        "groups": [
          "Errors"
        ],
        "analysisConfigurations": {
          "canary": {
            "nanStrategy": "replace",
            "direction": "decrease"
          }
        },
        "scopeName": "default"
      }
    ],
    "templates": {},
    "classifier": {
      "groupWeights": {
        "Errors": 34,
        "Saturation": 33,
        "Latency": 33
      }
    }
  }
}
```

Then the following queries are generated and sent to SignalFx for Bad Request Count

*BASELINE QUERY*
```
data('requests.count', filter=filter('response_code', '4*') and filter('server_group', 'carebears-cms-master-cerberus-prod2-baseline-v1')).sum(by=['server_group']).publish()
```

*CANARY QUERY*
```
data('requests.count', filter=filter('response_code', '4*') and filter('server_group', 'carebears-cms-master-cerberus-prod2-experiment-v1')).sum(by=['server_group']).publish()
```

It seems to me like you will need to reverse engineer the way Prometheus queries are generated and ensure that you are reporting your metrics in a manner that is consumable by Kayenta.

I would start at the [Query Builder](https://github.com/spinnaker/kayenta/blob/master/kayenta-prometheus/src/main/java/com/netflix/kayenta/prometheus/metrics/PrometheusMetricsService.java#L189) for Prometheus.

It also seems like the Prometheus metric source is adds the query to the metric set results https://github.com/spinnaker/kayenta/blob/master/kayenta-prometheus/src/main/java/com/netflix/kayenta/prometheus/metrics/PrometheusMetricsService.java#L305

I would take a look at one of those and check out the queries it made for you.

in the results from `standalone_canary_analysis`

![image](https://user-images.githubusercontent.com/711726/58494720-992d4080-812a-11e9-9b38-d9ba6fa9b12b.png)

Grab the `metricSetPairListId` for one of the results, then do a get on the following endpoint:

![image](https://user-images.githubusercontent.com/711726/58493869-bcef8700-8128-11e9-8b90-a2e89a83f116.png)

You will then be able to see the queries under attributes, like so (again this is for SignalFx, but this will work for your metric source)

![image](https://user-images.githubusercontent.com/711726/58494876-e7dada80-812a-11e9-82e4-038624ed097c.png)

**Comment by fieldju** [comment]
@spinnakerbot assign-issue @fieldju
